### ISSUE TITLE
GCN (global cycling network)

### ISSUE THREAD

**Issue by pete2583** [yt-dlp/yt-dlp#7324]
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting a new site support request
- [X] I've verified that I'm running yt-dlp version **2023.03.04** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that none of provided URLs [violate any copyrights](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy) or contain any [DRM](https://en.wikipedia.org/wiki/Digital_rights_management) to the best of my knowledge
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and am willing to share it if required

### Region

UK

### Example URLs

https://plus.globalcyclingnetwork.com/watch/1397691

### Provide a description that is worded well enough to be understood

Would love to be able to download videos from GCN offline. It requires a username and password

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
# yt-dlp https://plus.globalcyclingnetwork.com/watch/the-athertons-mountain-bikings-fastest-family --cookies --verbose
[generic] Extracting URL: https://plus.globalcyclingnetwork.com/watch/the-athertons-mountain-bikings-fastest-family
[generic] the-athertons-mountain-bikings-fastest-family: Downloading webpage
WARNING: [generic] Falling back on generic information extractor
[generic] the-athertons-mountain-bikings-fastest-family: Extracting information
ERROR: Unsupported URL: https://plus.globalcyclingnetwork.com/watch/the-athertons-mountain-bikings-fastest-family
```

**Comment by bashonly** [comment]
add verbose log with `--cookies-from-browser` or `--cookies` options being passed

**Comment by pete2583** [comment]
Edited to add verbose log

**Comment by bashonly** [comment]
You need to pass an argument with the `--cookies` option, e.g. `--cookies cookies.txt`

in the log you posted, yt-dlp interpreted `--cookies --verbose` as `--verbose` being the filename of your cookies file, and there was no verbose output (and no cookies were passed)

**Comment by pete2583** [comment]
Oh ok. That makes sense. 

Wouldn't have a clue how to pull the cookies from GCN though.

**Comment by pete2583** [comment]
Ok, 

I think I was able to figure the cookie thing out. So here's a verbose log (hopefully I've got it right)

`yt-dlp https://plus.globalcyclingnetwork.com/watch/1397691 --cookies cookie.txt --verbose
[debug] Command-line config: ['https://plus.globalcyclingnetwork.com/watch/1397691', '--cookies', 'cookie.txt', '--verbose']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version stable@2023.03.04 [392389b7d] (pip)
[debug] Python 3.10.4 (CPython x86_64 64bit) - Linux-5.15.90.1-microsoft-standard-WSL2-x86_64-with-glibc2.35 (OpenSSL 3.0.2 15 Mar 2022, glibc 2.35)
[debug] exe versions: ffmpeg 4.4.2 (setts), ffprobe 4.4.2
[debug] Optional libraries: Cryptodome-3.18.0, brotli-1.0.9, certifi-2020.06.20, mutagen-1.46.0, secretstorage-3.3.1, sqlite3-2.6.0, websockets-11.0.3
[debug] Proxy map: {}
[debug] Loaded 1786 extractors
[generic] Extracting URL: https://plus.globalcyclingnetwork.com/watch/1397691
[generic] 1397691: Downloading webpage
WARNING: [generic] Falling back on generic information extractor
[generic] 1397691: Extracting information
[debug] Looking for embeds
ERROR: Unsupported URL: https://plus.globalcyclingnetwork.com/watch/1397691
Traceback (most recent call last):
  File "/home/pete/.local/lib/python3.10/site-packages/yt_dlp/YoutubeDL.py", line 1518, in wrapper
    return func(self, *args, **kwargs)
  File "/home/pete/.local/lib/python3.10/site-packages/yt_dlp/YoutubeDL.py", line 1594, in __extract_info
    ie_result = ie.extract(url)
  File "/home/pete/.local/lib/python3.10/site-packages/yt_dlp/extractor/common.py", line 694, in extract
    ie_result = self._real_extract(url)
  File "/home/pete/.local/lib/python3.10/site-packages/yt_dlp/extractor/generic.py", line 2510, in _real_extract
    raise UnsupportedError(url)
yt_dlp.utils.UnsupportedError: Unsupported URL: https://plus.globalcyclingnetwork.com/watch/1397691`

**Comment by pete2583** [comment]
The trouble with GCN is I can skim the videos files through video downloader helper. 

It gives various video only downloads and then lists about 10-15 audio streams, all of which are about 5 seconds long so manually downloading them all and then stitching them together isn't really an option but I just can't figure out a way to get them to automatically download them all and stitch them together themselves.

**Comment by mpolom** [comment]
Sorry to reopen a dead issue, but GCN+ seems to now be DRM protected.

**Comment by qsevers23** [comment]
> Sorry to reopen a dead issue, but GCN+ seems to now be DRM protected.

Poor timing too with the announcement of [GCN+ ending in December too](https://www.globalcyclingnetwork.com/general/news/gcn-and-gcn-app-to-close)...

Anyone know if there's anything that can be fixed here? Or are we toast for downloading their films before it closes up?

**Comment by pukkandan** [comment]
We cannot support DRM content due to legal reasons

**Comment by mpolom** [comment]
> > Sorry to reopen a dead issue, but GCN+ seems to now be DRM protected.
> 
> Poor timing too with the announcement of [GCN+ ending in December](https://www.globalcyclingnetwork.com/general/news/gcn-and-gcn-app-to-close)...
> 
> Anyone know if there's anything that can be fixed here? Or are we toast for downloading their films before it closes up?

I suppose in that case it should be completely removed from the software?

**Comment by notjosh** [comment]
I haven't tried it extensively (yet!), and don't know if it's safe to apply across other dplus channels, but I noticed my browser (Firefox) wasn't getting DRM content from GCN.

The `streaming` object in the response in my browser looks like:

```json
{
        "cdn" : "akamai",
        "fallback" : false,
        "playbackProfile" : "h264-hls-ts-fhd-sdr-free-clear",
        "protection" : {
          "clearkeyEnabled" : false,
          "drmEnabled" : false,
          "schemes" : null
        },
        "provider" : "gps",
        "streamMode" : "VOD",
        "type" : "hls",
        "url" : "https://eurosport-vod.akamai.prod-live.h264.io/..."
      }
```

compared to this in `yt-dlp`'s existing implementation:

```json
{
        "cdn" : "akamai",
        "fallback" : false,
        "playbackProfile" : "h264-dash-fmp4-fhd-sdr-widevine-cenc",
        "protection" : {
          "clearkeyEnabled" : false,
          "drmEnabled" : true,
          "drmToken" : "[redacted]",
          "schemes" : {
            "widevine" : {
              "certificateUrl" : "https://discovery-eur.conax.cloud/widevine/certificates/conax",
              "licenseUrl" : "https://discovery-eur.conax.cloud/widevine/license"
            }
          }
        },
        "provider" : "gps",
        "streamMode" : "VOD",
        "type" : "dash",
        "url" : "https://eurosport-vod.akamai.prod-live.h264.io/..."
      }
```

Simply omitting the `wisteriaProperties` key from the request to `playback/v3/videoPlaybackInfo` can get `yt-dlp` to successfully pull down the video in 1080p. I don't _think_ there's any meaningful difference between the results (hls vs dash). If it's safe to make the change, can someone figure out how to test it across other properties, so we can save GCN content before it's gone?

tldr: GCN content isn't (all?) DRM'd, at least not in my region, by omitting a key from the request.

**Comment by notjosh** [comment]
> Happy to report that works for me too @notjosh. Not sure if that counts as circumventing DRM, but it works!

It's what my browser does by default (for some reason - maybe some privacy settings?), so I can only assume it's condoned.

**Comment by pukkandan** [comment]
Since the issues around this mostly comes down to difficulty testing, an implementation that uses the new method and then falls back to the old one on error would be best imo. We can issue a warning on fallback to users and see if anyone reports it, effectively crowd-sourcing the testing. If no reports come in for a while, that code branch can be removed.

**Comment by pukkandan** [comment]
> Not sure if that counts as circumventing DRM

No, it doesnt.

**Comment by bashonly** [comment]
> then falls back to the old one

the old method yields DRM-protected content 100% of the time though

**Comment by pukkandan** [comment]
> (#7530, #7955)

Then why are these needs-testing?

**Comment by bashonly** [comment]
I'm geo-blocked and/or don't have access to premium

**Comment by mpolom** [comment]
The GCN+ service gets turned off in 12 days, so although this might be useful for now it wouldn’t be in a short time!It would be interesting to see if this also happens with other Discovery content though - perhaps someone with another D+ subscription service can check.On 7 Dec 2023, at 17:23, bashonly ***@***.***> wrote:﻿
I'm geo-blocked and/or don't have access to premium

**Comment by qsevers23** [comment]
> I've created a plugin if anyone wants to help me test it: [yt-dlp-GlobalCyclingNetworkPlusDRMWorkaround](https://github.com/notjosh/yt-dlp-GlobalCyclingNetworkPlusDRMWorkaround).
> 
> After installing it, make sure to include the argument when calling `yt-dlp`. For example, I can run the following successfully:
> 
> ```shell
> $ yt-dlp -f bestaudio+bestvideo --cookies-from-browser firefox --extractor-args 'globalcyclingnetworkplus:use_drm_workaround' https://plus.globalcyclingnetwork.com/watch/5210311
> Extracting cookies from firefox
> Extracted 634 cookies from firefox
> [GlobalCyclingNetworkPlus+GlobalCyclingNetworkPlusDRMWorkaround] Extracting URL: https://plus.globalcyclingnetwork.com/watch/5210311
> [GlobalCyclingNetworkPlus+GlobalCyclingNetworkPlusDRMWorkaround] 5210311: Downloading JSON metadata
> [GlobalCyclingNetworkPlus+GlobalCyclingNetworkPlusDRMWorkaround] 5210311: Downloading JSON metadata
> [GlobalCyclingNetworkPlus+GlobalCyclingNetworkPlusDRMWorkaround] 5210311: Downloading m3u8 information
> [info] 5210311: Downloading 1 format(s): hls-audio_0-Default+hls-9005
> [hlsnative] Downloading m3u8 manifest
> [hlsnative] Total fragments: 597
> [download] Destination: UNBOUND [5210311].fhls-audio_0-Default.mp4
> [download] 100% of   54.63MiB in 00:00:43 at 1.25MiB/s
> [hlsnative] Downloading m3u8 manifest
> [hlsnative] Total fragments: 597
> [download] Destination: UNBOUND [5210311].fhls-9005.mp4
> [download] 100% of    3.29GiB in 00:02:47 at 20.09MiB/s
> [Merger] Merging formats into "UNBOUND [5210311].mp4"
> Deleting original file UNBOUND [5210311].fhls-9005.mp4 (pass -k to keep)
> Deleting original file UNBOUND [5210311].fhls-audio_0-Default.mp4 (pass -k to keep)
> 
> $ ls -l ./UNBOUND\ \[5210311\].mp4
> .rw-r--r--@ 3.5G user group 2023-12-07 16:05 ./'UNBOUND [5210311].mp4'
> ```
> 
> (...it's my first `yt-dlp` plugin, so let me know if there are any problems!)

Having some issues with this, still replying back with DRM protected

```
➜  gcn yt-dlp -f bestaudio+bestvideo --cookies-from-browser firefox --extractor-args 'globalcyclingnetworkplus:use_drm_workaround' https://plus.globalcyclingnetwork.com/watch/5210311
Extracting cookies from firefox
Extracted 47 cookies from firefox
[GlobalCyclingNetworkPlus] Extracting URL: https://plus.globalcyclingnetwork.com/watch/5210311
[GlobalCyclingNetworkPlus] 5210311: Downloading JSON metadata
[GlobalCyclingNetworkPlus] 5210311: Downloading JSON metadata
[GlobalCyclingNetworkPlus] 5210311: Downloading MPD manifest
ERROR: [GlobalCyclingNetworkPlus] 5210311: This video is DRM protected
```

Maybe a region issue (quite possibly user error as well)? I'm in US.
